src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -2172,6 +2172,50 @@ private SafeContext GetInvocationEscapeToReceiver(
             return escapeScope;
         }
 
+        private static MethodInfo ReplaceWithExtensionImplementationIfNeeded(MethodInfo methodInfo, ref ImmutableArray<ParameterSymbol> parameters,
+            ref BoundExpression? receiver, ref ImmutableArray<BoundExpression> argsOpt, ref ImmutableArray<RefKind> argRefKindsOpt,
+            ref ImmutableArray<int> argsToParamsOpt)
+        {
+            Symbol? symbol = methodInfo.Symbol;
+            if (symbol?.GetIsNewExtensionMember() != true || symbol.IsStatic)
+            {
+                return methodInfo;
+            }
+
+            MethodInfo replacedMethodInfo = methodInfo.ReplaceWithExtensionImplementation(out bool wasError);
+            if (wasError)
+            {
+                return methodInfo;
+            }
+
+            var extensionParameter = symbol.ContainingType.ExtensionParameter;
+            Debug.Assert(extensionParameter is not null);
+            parameters = parameters.IsDefault ? [extensionParameter] : [extensionParameter, .. parameters];
+
+            Debug.Assert(receiver is not null);
+            argsOpt = argsOpt.IsDefault ? [receiver] : [receiver, .. argsOpt];
+            receiver = null;
+
+            if (!argRefKindsOpt.IsDefault)
+            {
+                argRefKindsOpt = [RefKind.None, .. argRefKindsOpt];
```

Comment:
I still need to review the larger PR, but, I am confused by this part. Doesn't the receiver argument ref kind need to be compatible with the extension parameter? e.g. `extension(ref int i)` needs a compatible ref kind here.

Reply:
😓